File: platforms/apple/Sources/BotaDeviceSDK/FactoryResetManager.swift
import Foundation

public struct FactoryResetGrantRequest: Sendable {
    public let serialNumber: String
    public let nonce: Data
    public let commandID: String
    public let bindingGeneration: UInt64

    public init(serialNumber: String, nonce: Data, commandID: String, bindingGeneration: UInt64) {
        self.serialNumber = serialNumber
        self.nonce = nonce
        self.commandID = commandID
        self.bindingGeneration = bindingGeneration
    }
}

public typealias FactoryResetGrantProvider = @Sendable (FactoryResetGrantRequest) async throws -> Data

public struct FactoryResetCompletion: Equatable, Sendable {
    public let commandID: String
    public let bindingGeneration: UInt64

    public init(commandID: String, bindingGeneration: UInt64) {
        self.commandID = commandID
        self.bindingGeneration = bindingGeneration
    }
}

public actor FactoryResetManager {
    private var runtime: DeviceRuntime?
    private var activeCancellationID: UUID?

    public init() {}

    func attach(_ runtime: DeviceRuntime) { self.runtime = runtime }

    func detach() async {
        if let activeCancellationID {
            try? await runtime?.engine.cancel(activeCancellationID)
            await runtime?.operations.end(activeCancellationID)
        }
        activeCancellationID = nil
        runtime = nil
    }

    public func factoryReset(
        _ device: ConnectedDevice,
        commandID: String,
        grantID: String,
        bindingGeneration: UInt64,
        using provider: @escaping FactoryResetGrantProvider
    ) async throws -> FactoryResetCompletion {
        let runtime = try configuredRuntime()
        try await runtime.connection.require(device)
        await runtime.registerFactoryResetGeneration(commandID, bindingGeneration)
        await runtime.registerFactoryReset(grantID) { request in
            try await provider(.init(
                serialNumber: request.serialNumber,
                nonce: request.nonce,
                commandID: commandID,
                bindingGeneration: bindingGeneration
            ))
        }
        do {
            try await run(
                .factoryReset(
                    serialNumber: device.serialNumber,
                    commandID: commandID,
                    grantID: grantID
                ),
                runtime: runtime
            )
            await cleanup(commandID: commandID, grantID: grantID, runtime: runtime)
            return FactoryResetCompletion(
                commandID: commandID,
                bindingGeneration: bindingGeneration
            )
        } catch {
            await cleanup(commandID: commandID, grantID: grantID, runtime: runtime)
            throw error
        }
    }

    public func resumePendingFactoryReset(
        _ device: ConnectedDevice,
        currentBindingGeneration: UInt64
    ) async throws -> FactoryResetCompletion? {
        let runtime = try configuredRuntime()
        try await runtime.connection.require(device)
        guard let saved = try await runtime.loadPendingFactoryReset() else { return nil }
        guard saved.bindingGeneration == currentBindingGeneration else {
            throw BotaDeviceSDKError(
                code: .identityMismatch,
                operation: .factoryReset,
                retryable: false,
                detail: "pending factory reset belongs to a different binding generation"
            )
        }
        guard let resultCode = UInt8(exactly: saved.resultCode),
              let deletedCount = UInt16(exactly: saved.deletedRecordingCount)
        else {
            throw BotaDeviceSDKError(
                code: .persistenceFailed,
                operation: .factoryReset,
                retryable: false,
                detail: "pending factory-reset result is out of range"
            )
        }
        try await run(
            .resumeFactoryReset(
                serialNumber: device.serialNumber,
                commandID: saved.commandID,
                resultCode: resultCode,
                deletedRecordingCount: deletedCount
            ),
            runtime: runtime
        )
        return FactoryResetCompletion(
            commandID: saved.commandID,
            bindingGeneration: currentBindingGeneration
        )
    }

    public func cancelCurrentOperation() async throws {
        guard let activeCancellationID else { return }
        let runtime = try configuredRuntime()
        try await runtime.engine.cancel(activeCancellationID)
        self.activeCancellationID = nil
        await runtime.operations.end(activeCancellationID)
    }

    private func run(_ command: CoreCommand, runtime: DeviceRuntime) async throws {
        guard activeCancellationID == nil else {
            throw BotaDeviceSDKError(
                code: .operationInProgress,
                operation: .factoryReset,
                retryable: false,
                detail: "another factory-reset workflow is already active"
            )
        }
        try await runtime.operations.begin(command.cancellationID, operation: .factoryReset)
        activeCancellationID = command.cancellationID
        do {
            try await awaitWorkflowCompletion(command, runtime: runtime)
            activeCancellationID = nil
            await runtime.operations.end(command.cancellationID)
        } catch {
            activeCancellationID = nil
            await runtime.operations.end(command.cancellationID)
            throw error
        }
    }

    private func cleanup(commandID: String, grantID: String, runtime: DeviceRuntime) async {
        await runtime.unregisterMaterial(grantID)
        await runtime.unregisterFactoryResetGeneration(commandID)
    }

    private func configuredRuntime() throws -> DeviceRuntime {
        guard let runtime else {
            throw BotaDeviceSDKError(
                code: .featureUnavailable,
                operation: .validate,
                retryable: false,
                detail: "BotaDeviceClient.configure() must be called first"
            )
        }
        return runtime
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/FactoryResetManager.kt
package dev.bota.sdk

import dev.bota.sdk.internal.DeviceRuntime
import dev.bota.sdk.internal.awaitWorkflowCompletion
import dev.bota.sdk.internal.core.CoreCommand
import dev.bota.sdk.internal.runCleanupActions
import dev.bota.sdk.internal.runCleanupAfter
import dev.bota.sdk.model.ConnectedDevice
import dev.bota.sdk.model.FactoryResetCompletion
import dev.bota.sdk.model.FactoryResetGrantRequest
import dev.bota.sdk.model.FactoryResetPersistenceResult
import java.util.UUID

public class FactoryResetManager internal constructor() {
    private data class Active(
        val cancellationId: UUID,
        val commandId: String,
        val grantId: String?,
    )

    private val lock = Any()
    private var runtime: DeviceRuntime? = null
    private var active: Active? = null

    internal fun attach(runtime: DeviceRuntime) {
        synchronized(lock) { this.runtime = runtime }
    }

    internal suspend fun detach() {
        val snapshot = synchronized(lock) {
            (runtime to active).also {
                runtime = null
                active = null
            }
        }
        val configured = snapshot.first ?: return
        snapshot.second?.let { operation ->
            runCleanupActions(
                { runCatching { configured.engine.cancel(operation.cancellationId) } },
                { operation.grantId?.let(configured.unregisterMaterial) },
                { configured.unregisterFactoryResetResultPersister(operation.commandId) },
                { configured.unregisterFactoryResetGeneration(operation.commandId) },
                { configured.operations.end(operation.cancellationId) },
            )
        }
    }

    public suspend fun factoryReset(
        device: ConnectedDevice,
        commandId: String,
        bindingGeneration: ULong,
        persistResult: (suspend (FactoryResetPersistenceResult) -> Unit)? = null,
        provider: suspend (FactoryResetGrantRequest) -> ByteArray,
    ): FactoryResetCompletion {
        requireIdentifier(commandId, "command ID")
        val configured = configuredRuntime()
        configured.connection.require(device)
        val grantId = UUID.randomUUID().toString()
        val command = CoreCommand.factoryReset(device.serialNumber, commandId, grantId)
        configured.operations.begin(command.cancellationId, BotaOperation.FactoryReset)
        synchronized(lock) { active = Active(command.cancellationId, commandId, grantId) }
        var failure: Throwable? = null
        try {
            configured.registerFactoryResetGeneration(commandId, bindingGeneration)
            persistResult?.let { persister ->
                configured.registerFactoryResetResultPersister(commandId) { result ->
                    persister(result.toPublicPersistenceResult())
                }
            }
            configured.registerFactoryReset(grantId) { serialNumber, nonce ->
                provider(FactoryResetGrantRequest(serialNumber, nonce, commandId, bindingGeneration))
            }
            awaitWorkflowCompletion(command, configured)
            return FactoryResetCompletion(commandId, bindingGeneration)
        } catch (error: Throwable) {
            failure = error
            throw error
        } finally {
            runCleanupAfter(
                failure,
                { configured.unregisterMaterial(grantId) },
                { configured.unregisterFactoryResetResultPersister(commandId) },
                { configured.unregisterFactoryResetGeneration(commandId) },
                { finish(command.cancellationId) },
            )
        }
    }

    public suspend fun resumePendingFactoryReset(
        device: ConnectedDevice,
        currentBindingGeneration: ULong,
        persistResult: (suspend (FactoryResetPersistenceResult) -> Unit)? = null,
    ): FactoryResetCompletion? {
        val configured = configuredRuntime()
        configured.connection.require(device)
        val saved = configured.loadPendingFactoryReset() ?: return null
        if (saved.bindingGeneration != currentBindingGeneration) throw BotaSDKError.Core(
            BotaErrorCode.IdentityMismatch,
            BotaOperation.FactoryReset,
            retryable = false,
            protocolStatus = null,
            detail = "pending factory reset belongs to a different binding generation",
        )
        if (saved.resultCode > UByte.MAX_VALUE.toULong() ||
            saved.deletedRecordingCount > UShort.MAX_VALUE.toULong()
        ) {
            throw BotaSDKError.Core(
                BotaErrorCode.PersistenceFailed,
                BotaOperation.FactoryReset,
                retryable = false,
                protocolStatus = null,
                detail = "pending factory-reset result is out of range",
            )
        }
        val command = CoreCommand.resumeFactoryReset(
            device.serialNumber,
            saved.commandId,
            saved.resultCode.toUByte(),
            saved.deletedRecordingCount.toUShort(),
        )
        configured.operations.begin(command.cancellationId, BotaOperation.FactoryReset)
        synchronized(lock) { active = Active(command.cancellationId, saved.commandId, null) }
        configured.registerFactoryResetGeneration(saved.commandId, currentBindingGeneration)
        persistResult?.let { persister ->
            configured.registerFactoryResetResultPersister(saved.commandId) { result ->
                persister(result.toPublicPersistenceResult())
            }
        }
        try {
            awaitWorkflowCompletion(command, configured)
            return FactoryResetCompletion(saved.commandId, currentBindingGeneration)
        } finally {
            runCleanupActions(
                { configured.unregisterFactoryResetResultPersister(saved.commandId) },
                { configured.unregisterFactoryResetGeneration(saved.commandId) },
                { finish(command.cancellationId) },
            )
        }
    }

    /**
     * Resumes a firmware-held reset result when this installation has no native
     * journal. No grant or reset opcode is sent.
     */
    public suspend fun resumeUnjournaledFactoryReset(
        device: ConnectedDevice,
        commandId: String,
        bindingGeneration: ULong,
        persistResult: suspend (FactoryResetPersistenceResult) -> Unit,
    ): FactoryResetCompletion {
        requireIdentifier(commandId, "command ID")
        val configured = configuredRuntime()
        configured.connection.require(device)
        val command = CoreCommand.resumeUnjournaledFactoryReset(device.serialNumber, commandId)
        configured.operations.begin(command.cancellationId, BotaOperation.FactoryReset)
        synchronized(lock) { active = Active(command.cancellationId, commandId, null) }
        configured.registerFactoryResetGeneration(commandId, bindingGeneration)
        configured.registerFactoryResetResultPersister(commandId) { result ->
            persistResult(result.toPublicPersistenceResult())
        }
        try {
            awaitWorkflowCompletion(command, configured)
            return FactoryResetCompletion(commandId, bindingGeneration)
        } finally {
            runCleanupActions(
                { configured.unregisterFactoryResetResultPersister(commandId) },
                { configured.unregisterFactoryResetGeneration(commandId) },
                { finish(command.cancellationId) },
            )
        }
    }

    public suspend fun cancelCurrentOperation() {
        val configured = configuredRuntime()
        val operation = synchronized(lock) { active } ?: return
        configured.engine.cancel(operation.cancellationId)
    }

    private fun finish(cancellationId: UUID) {
        val configured = synchronized(lock) {
            if (active?.cancellationId != cancellationId) return
            active = null
            runtime
        }
        configured?.operations?.end(cancellationId)
    }

    private fun configuredRuntime(): DeviceRuntime = synchronized(lock) { runtime } ?: unavailable()
}

private fun dev.bota.sdk.internal.host.PersistedFactoryResetResult.toPublicPersistenceResult():
    FactoryResetPersistenceResult {
    if (resultCode != 0uL || deletedRecordingCount > UShort.MAX_VALUE.toULong()) {
        throw BotaSDKError.Core(
            BotaErrorCode.PersistenceFailed,
            BotaOperation.FactoryReset,
            retryable = false,
            protocolStatus = null,
            detail = "factory-reset persistence result is out of range",
        )
    }
    return FactoryResetPersistenceResult(deletedRecordingCount.toUShort())
}

internal object SecureUUIDs {
    val ControlService: UUID = UUID.fromString("b07a0002-0000-1000-8000-00805f9b34fb")
    val DeviceCommand: UUID = UUID.fromString("b07a0002-0005-1000-8000-00805f9b34fb")
    val ProvisioningService: UUID = UUID.fromString("b07a0003-0000-1000-8000-00805f9b34fb")
    val ProvisioningResult: UUID = UUID.fromString("b07a0003-0002-1000-8000-00805f9b34fb")
    val DeviceSettings: UUID = UUID.fromString("b07a0003-0006-1000-8000-00805f9b34fb")
}

private fun requireIdentifier(value: String, label: String) {
    if (value.isBlank()) throw BotaSDKError.Core(
        BotaErrorCode.InvalidInput,
        BotaOperation.Validate,
        retryable = false,
        protocolStatus = null,
        detail = "$label is required",
    )
}

internal fun unavailable(): Nothing = throw BotaSDKError.Core(
    BotaErrorCode.FeatureUnavailable,
    BotaOperation.Validate,
    retryable = false,
    protocolStatus = null,
    detail = "BotaDeviceClient.configure() must be called first",
)
